kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs

```diff
@@ -249,10 +262,20 @@ impl<'tcx> GotocCtx<'tcx> {
                 let (tmp, decl_stmt) =
                     self.decl_temp_variable(var1.typ().clone(), Some(var1.to_owned()), loc);
                 let var2 = fargs.remove(0);
-                let op_expr = (var1.clone()).$op(var2).with_location(loc);
-                let assign_stmt = (var1.clone()).assign(op_expr, loc);
-                let res_stmt = self.codegen_expr_to_place_stable(place, tmp.clone());
-                Stmt::atomic_block(vec![decl_stmt, assign_stmt, res_stmt], loc)
+                if var2.typ().is_pointer() {
+                    let op_expr = (var1.clone().cast_to(Type::c_size_t()))
```

Comment:
Are the casts here necessary, or is the last one (casting the whole expression to `var1.typ()`) sufficient?

Reply:
Yes, they are necessary. For example, in https://doc.rust-lang.org/std/sync/atomic/struct.AtomicPtr.html#method.fetch_byte_add, we do pointer arithmetic `*var1-var2` in the intrinsic where `var2` is a pointer `invalid_mut`. It will lead to typecheck failure since we don't allow `pointer - pointer`. Actually, the hack in `fetch_byte_add` is to use `invalid_mut` as integer to do the pointer arithmetic. So we also cast the variables to integer type in the codegen macro to avoid type check failures.